docs/solrbulk.1
.TH SOLRBULK 1 "JANUAR 2015" "Leipzig University Library" "Manuals"
.SH NAME
.PP
solrbulk \- send documents to SOLR in bulk and in parallel
.SH SYNOPSIS
.PP
\fB\fCsolrbulk\fR [\fB\fC\-server\fR \fIURL\fP, \fB\fC\-size\fR \fIN\fP, \fB\fC\-w\fR \fIN\fP, \fB\fC\-commit\fR \fIN\fP, \fB\fC\-z\fR] \fIfile\fP
.SH DESCRIPTION
.PP
solrbulk takes as input a newline delimited JSON file and indexes all documents
into SOLR running on a given server address. The documents are batched and
indexed in parallel to achieve a high indexing throughput.
.SH OPTIONS
.TP
\fB\fC\-auth\fR \fIcred\fP
username:password pair for basic auth
.TP
\fB\fC\-commit\fR \fIN\fP
Issue a commit every N documents.
.TP
\fB\fC\-cpuprofile\fR \fIfilename\fP
Write cpu profile to given filename.
.TP
\fB\fC\-memprofile\fR \fIfilename\fP
Write memory profile to given filename.
.TP
\fB\fC\-no\-final\-commit\fR
Omit final commit.
.TP
\fB\fC\-optimize\fR
Issue an optimize after final commit.
.TP
\fB\fC\-purge\fR
Remove documents from index before indexing (use purge\-query to selectively clean). No questions asked.
.TP
\fB\fC\-purge\-pause\fR \fIduration\fP
Insert a short pause after a purge query, defaults to 2s.
.TP
\fB\fC\-purge\-query\fR \fIquery\fP
Query to use, when purging (default "*:*").
.TP
\fB\fC\-server\fR \fIURL\fP
SOLR URL including host, port and core, like \[la]http://localhost:8983/solr/biblio\[ra]\&. To change the path of the update request handler, use \fB\fC\-update\-request\-handler\-name\fR flag.
.TP
\fB\fC\-size\fR \fIN\fP
Batch size.
.TP
\fB\fC\-update\-request\-handler\-name\fR \fIname\fP
Where solr.UpdateRequestHandler is mounted on the server, \[la]https://is.gd/s0eirv\[ra] (default "/update").
.TP
\fB\fC\-v\fR
Program version.
.TP
\fB\fC\-verbose\fR
Show progress.
.TP
\fB\fC\-w\fR \fIN\fP
Number of workers.
.TP
\fB\fC\-z\fR
Decompress gzip input file on the fly.
.SH EXAMPLES
.PP
Index a file:
.IP
\fB\fCsolrbulk \-verbose \-server 110.81.131.200:8080/solr/biblio file.ldj\fR
.PP
Index from compressed standard input:
.IP
\fB\fCsolrbulk \-z \-server 110.81.131.200:8080/solr/biblio < file.ndj.gz\fR
.PP
Index a file, but remove all docs from the index before indexing:
.IP
\fB\fCsolrbulk \-purge 110.81.131.200:8080/solr/biblio file.ldj\fR
.SH LIMITATIONS
.PP
If more than one worker is used for indexing, the order in which documents get
indexed does not need to equal the order in which they appear in originating
stream. In the best case, indexing order should not matter at all.
.SH PERFORMANCE
.PP
Real\-world performance data point: Indexing 150M 2K documents on a single
server with 8+ cores and ample RAM runs at sustained 5K docs/s.
.SH BUGS
.PP
Please report bugs to \[la]https://github.com/miku/solrbulk/issues\[ra]\&.
.SH AUTHORS
.PP
Martin Czygan \[la]martin.czygan@uni-leipzig.de\[ra], Thomas Gersch \[la]https://github.com/titabo2k\[ra]
.SH SEE ALSO
.PP
FINC \[la]https://finc.info\[ra], AMSL \[la]http://amsl.technology/\[ra], esbulk \[la]https://github.com/miku/esbulk\[ra]


cmd/solrbulk/solrbulk.go
// solrbulk sends documents to a SOLR server.
//
//  Copyright 2015 by Leipzig University Library, http://ub.uni-leipzig.de
//                 by The Finc Authors, http://finc.info
//                 by Martin Czygan, <martin.czygan@uni-leipzig.de>
//
// This file is part of some open source application.
//
// Some open source application is free software: you can redistribute
// it and/or modify it under the terms of the GNU General Public
// License as published by the Free Software Foundation, either
// version 3 of the License, or (at your option) any later version.
//
// Some open source application is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
//
// @license GPL-3.0+ <http://spdx.org/licenses/GPL-3.0+>

package main

import (
	"bufio"
	b64 "encoding/base64"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"runtime"
	"runtime/pprof"
	"strings"
	"sync"
	"time"

	gzip "github.com/klauspost/compress/gzip"
	"github.com/miku/solrbulk"
	"github.com/sethgrid/pester"
	log "github.com/sirupsen/logrus"
)

var (
	version                  = flag.Bool("v", false, "prints current program version")
	cpuprofile               = flag.String("cpuprofile", "", "write cpu profile to file")
	memprofile               = flag.String("memprofile", "", "write heap profile to file")
	batchSize                = flag.Int("size", 1000, "bulk batch size")
	commitSize               = flag.Int("commit", 1000000, "commit after this many docs")
	numWorkers               = flag.Int("w", runtime.NumCPU(), "number of workers to use")
	verbose                  = flag.Bool("verbose", false, "output basic progress")
	gzipped                  = flag.Bool("z", false, "unzip gz'd file on the fly")
	server                   = flag.String("server", "", "url to SOLR server, including host, port and path to collection, e.g. http://localhost:8983/solr/biblio")
	optimize                 = flag.Bool("optimize", false, "optimize index")
	purge                    = flag.Bool("purge", false, "remove documents from index before indexing (use purge-query to selectively clean)")
	purgeQuery               = flag.String("purge-query", "*:*", "query to use, when purging")
	purgePause               = flag.Duration("purge-pause", 2*time.Second, "insert a short pause after purge")
	updateRequestHandlerName = flag.String("update-request-handler-name", "/update", "where solr.UpdateRequestHandler is mounted on the server, https://is.gd/s0eirv")
	noFinalCommit            = flag.Bool("no-final-commit", false, "omit final commit")
	basicAuth                = flag.String("auth", "", "username:password pair for basic auth")
)

func newGetRequest(url string, options solrbulk.Options) (*http.Request, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	if options.BasicAuth != "" {
		req.Header.Add("Authorization", "Basic "+b64.StdEncoding.EncodeToString([]byte(options.BasicAuth)))
	}
	return req, nil
}

func main() {
	flag.Parse()
	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		if err := pprof.StartCPUProfile(f); err != nil {
			log.Fatal(err)
		}
		defer pprof.StopCPUProfile()
	}
	if *version {
		fmt.Println(solrbulk.Version)
		os.Exit(0)
	}
	options := solrbulk.Options{
		BatchSize:                *batchSize,
		CommitSize:               *commitSize,
		Verbose:                  *verbose,
		UpdateRequestHandlerName: *updateRequestHandlerName,
		Server:                   *server,
		BasicAuth:                *basicAuth,
	}
	if !strings.HasPrefix(options.Server, "http") {
		options.Server = fmt.Sprintf("http://%s", options.Server)
	}
	if *purge {
		var (
			hostpath = fmt.Sprintf("%s%s", options.Server, options.UpdateRequestHandlerName)
			urls     = []string{
				fmt.Sprintf("%s?stream.body=<delete><query>%s</query></delete>", hostpath, *purgeQuery),
				fmt.Sprintf("%s?stream.body=<commit/>", hostpath),
			}
		)
		for _, url := range urls {
			req, err := newGetRequest(url, options)
			if err != nil {
				log.Fatal(err)
			}

			resp, err := pester.Do(req)
			if err != nil {
				log.Fatal(err)
			}
			log.Printf("%s %s", resp.Status, url)
		}
		time.Sleep(*purgePause)
	}
	var file io.Reader = os.Stdin
	if flag.NArg() > 0 {
		f, err := os.Open(flag.Arg(0))
		if err != nil {
			log.Fatalln(err)
		}
		defer f.Close()
		file = f
	}
	var (
		queue     = make(chan string)
		wg        sync.WaitGroup
		commitURL = fmt.Sprintf("%s%s?commit=true", options.Server, options.UpdateRequestHandlerName)
		reader    = bufio.NewReader(file)
	)
	for i := 0; i < *numWorkers; i++ {
		wg.Add(1)
		go solrbulk.Worker(fmt.Sprintf("worker-%d", i), options, queue, &wg)
	}
	if !*noFinalCommit {
		defer func() {
			resp, err := pester.Get(commitURL)
			if err != nil {
				log.Fatal(err)
			}
			if err := resp.Body.Close(); err != nil {
				log.Fatal(err)
			}
			log.Printf("final commit: %s", resp.Status)
		}()
	}
	if *gzipped {
		zreader, err := gzip.NewReader(reader)
		if err != nil {
			log.Fatal(err)
		}
		reader = bufio.NewReader(zreader)
	}
	var (
		i     = 0
		start = time.Now()
	)
	for {
		line, err := reader.ReadString('\n')
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		line = strings.TrimSpace(line)
		queue <- line
		i++
		if i%options.CommitSize == 0 {
			req, err := newGetRequest(commitURL, options)
			if err != nil {
				log.Fatal(err)
			}

			resp, err := pester.Do(req)
			if err != nil {
				log.Fatal(err)
			}
			if options.Verbose {
				log.Printf("commit @%d %s", i, resp.Status)
			}
			if err := resp.Body.Close(); err != nil {
				log.Fatal(err)
			}
		}
	}
	close(queue)
	wg.Wait()
	if *memprofile != "" {
		f, err := os.Create(*memprofile)
		if err != nil {
			log.Fatal(err)
		}
		if err := pprof.WriteHeapProfile(f); err != nil {
			log.Fatal(err)
		}
		f.Close()
	}
	if *verbose {
		elapsed := time.Since(start)
		rate := float64(i) / elapsed.Seconds()
		log.Printf("%d docs in %s at %0.3f docs/s with %d workers", i, elapsed, rate, *numWorkers)
	}
	if *optimize {
		hostpath := fmt.Sprintf("%s%s", options.Server, options.UpdateRequestHandlerName)
		url := fmt.Sprintf("%s?stream.body=<optimize/>", hostpath)

		req, err := newGetRequest(url, options)
		if err != nil {
			log.Fatal(err)
		}

		resp, err := pester.Do(req)
		if err != nil {
			log.Fatal(err)
		}
		log.Printf("%s %s", resp.Status, url)
		elapsed := time.Since(start)
		if *verbose {
			log.Printf("indexed and optimized in %s", elapsed)
		}
	}
}
